### PR TITLE
Fix/fiducial rng

### PR DESCRIPTION
**Pull request by johnkgamble**
The GRASP implementation of fiducial generation was ignoring the `seed` kwarg passed to the solver. This PR addresses the issue by instancing a random number generator with a given seed in GRASP-based fiducial generation. This rng is then passed around as needed by GRASP. 

Note: many of the changes here are just black auto-formatting. The substantive changes are quite minimal.

### CODE REVIEW

File: pygsti/algorithms/fiducialselection.py [hunk 1]
```diff
@@ -945,14 +1039,22 @@ def _find_fiducials_grasp(model, fids_list, prep_or_meas, alpha,
...
     printer = _baseobjs.VerbosityPrinter.create_printer(verbosity)
-
-    if prep_or_meas not in ['prep', 'meas']:
-        raise ValueError("'{}' is an invalid value for prep_or_meas (must be "
-                         "'prep' or 'meas')!".format(prep_or_meas))
-
-    initial_test = test_fiducial_list(model, fids_list, prep_or_meas,
-                                      score_func=score_func, return_all=False,
-                                      threshold=threshold)
+    rng = random.Random(seed)
```
Comment (enielse) [inline pygsti/algorithms/fiducialselection.py:1042]: Marking this as a substantial (non-style) change

File: pygsti/algorithms/fiducialselection.py [hunk 2]
```diff
@@ -966,73 +1068,77 @@ def _find_fiducials_grasp(model, fids_list, prep_or_meas, alpha,
...
+                    elements=fids_list,
+                    greedy_score_fn=score_fn,
+                    rcl_fn=rcl_fn,
                     local_score_fn=score_fn,
                     get_neighbors_fn=_get_neighbors_fn,
                     feasible_threshold=feasibleThreshold,
-                    initial_elements=initialWeights, seed=seed,
-                    verbosity=verbosity)
+                    initial_elements=initialWeights,
+                    rng=rng,
```
Comment (enielse) [inline pygsti/algorithms/fiducialselection.py:1123]: Marked as a substantial change.

File: pygsti/algorithms/grasp.py [hunk 3]
```diff
@@ -67,8 +66,15 @@ def neighboring_weight_vectors(weights, forced_weights=None, shuffle=False):
...
-def _grasp_construct_feasible_solution(elements, score_fn, rcl_fn, feasible_threshold=None,
-                                       feasible_fn=None, initial_elements=None):
+def _grasp_construct_feasible_solution(
+    elements,
+    score_fn,
+    rcl_fn,
+    feasible_threshold=None,
+    feasible_fn=None,
+    initial_elements=None,
+    rng=None,
```
Comment (enielse) [inline pygsti/algorithms/grasp.py:76]: Marked as substantial change - adds rng arg to signature, docstring, and within function body below.

File: pygsti/algorithms/grasp.py [hunk 4]
```diff
@@ -118,56 +124,72 @@ def _grasp_construct_feasible_solution(elements, score_fn, rcl_fn, feasible_thre
...
-        candidateScores = _np.array([score_fn(candidateSoln)
-                                     for candidateSoln in candidateSolns])
+        candidateScores = _np.array(
+            [score_fn(candidateSoln) for candidateSoln in candidateSolns]
+        )
         rclIdxs = rcl_fn(candidateScores)
-        assert(len(rclIdxs) > 0), "Empty reduced candidate list!"
-        chosenIdx = _np.random.choice(rclIdxs)
+        assert len(rclIdxs) > 0, "Empty reduced candidate list!"
+        chosenIdx = rng.choice(rclIdxs)
```
Comment (enielse) [inline pygsti/algorithms/grasp.py:171]: This is a substantial change, in particular

File: pygsti/algorithms/grasp.py [hunk 5]
```diff
@@ -267,9 +293,18 @@ def _grasp_local_search(initial_solution, score_fn, elements, get_neighbors_fn,
...
+def run_grasp_iteration(
+    elements,
+    greedy_score_fn,
+    rcl_fn,
+    local_score_fn,
+    get_neighbors_fn,
+    feasible_threshold=None,
+    feasible_fn=None,
+    initial_elements=None,
+    rng=None,
```
Comment (enielse) [inline pygsti/algorithms/grasp.py:305]: Here's another substantial change - adding `rng` argument again in place of the `seed` argument.

File: pygsti/algorithms/grasp.py [hunk 6]
```diff
@@ -427,15 +489,21 @@ def run_grasp(elements, greedy_score_fn, rcl_fn, local_score_fn, get_neighbors_f
     printer = _baseobjs.VerbosityPrinter.create_printer(verbosity)
 
     bestSoln = None
+    rng = random.Random(seed)
```
Comment (enielse) [inline pygsti/algorithms/grasp.py:492]: Marked as substantial change

File: pygsti/algorithms/grasp.py [hunk 7]
```diff
@@ -427,15 +489,21 @@ def run_grasp(elements, greedy_score_fn, rcl_fn, local_score_fn, get_neighbors_f
...
+        _, localSoln = run_grasp_iteration(
+            elements,
+            greedy_score_fn,
+            rcl_fn,
+            local_score_fn,
+            get_neighbors_fn,
+            feasible_threshold=feasible_threshold,
+            feasible_fn=feasible_fn,
+            initial_elements=initial_elements,
+            rng=rng,
```
Comment (enielse) [inline pygsti/algorithms/grasp.py:504]: Marked as substantial change

**Review by enielse — CHANGES_REQUESTED** [review]
Technically this PR seems fine, and adds plumbing to send a random number generator down into the GRASP algorithm internals so that the randomness can be replicated deterministically using a (pseudorandom) number generator.  However, most of the ~650 changes are just stylistic (e.g. whitespace conventions, single vs double quotes, etc.) and don't change functionality at all.  (Maybe an editing tool was used that does this automatically according to its own Python style preferences?)  Before accepting this PR, lets clean this up so that only the substantive changes are made, as this makes tracking changes much easier and cleaner.  I've gone through and marked what I think are all the substantive changes in my review comments.  Assuming these are correct, I'm happy to create a separate PR that just updates these parts.

**Comment by johnkgamble** [comment]
@enielse - correct you are, sorry! Most of the changes are due to [black](https://black.readthedocs.io/en/stable/). I'll push up a minimal version later today. 

Black is extremely useful overall and used by tons of projects; I'm going to open an issue ticket to discuss the prospect of running all of pygsti through it to make everything easier.

**Comment by enielse** [comment]
Ah, right - this was mentioned in the original PR but I didn't realize "black" was an editor :).  Yes, we'd be open to bulk reformatting but I think it's wise to keep that separated from functional changes to keep the history more readable.  Starting a separate discussion in an issue seems like a great plan.

**Comment by johnkgamble** [comment]
@enielse , I've made just the minimal changes now - can you take another look? Thanks!

**Review by enielse — APPROVED** [review]
Looks great!

**Pull request merged**